### PR TITLE
Add factory SPI for `@TempDir`

### PR DESCRIPTION
**Pull request by scordio**
## TL;DR

This could solve:

* #2088
* #2400
* #2889
* #3168

## Overview

This adds a factory SPI to `@TempDir`, allowing to define how the temporary directory is created.

My initial target was to support a custom file system (#2400), but it may also be a solution for the issues mentioned above.

The intended usage is to extend `TempDirFactory` and provide the class to the new `factory` attribute of `@TempDir`:

```java
...
import java.nio.file.Files;
...

class CustomFactory implements TempDirFactory {

    @Override
    public Path createTempDirectory(ExtensionContext context) throws IOException {
        return Files.createTempDirectory("junit");
    }

}
```

```java
@Test
void test(@TempDir(factory = CustomFactory.class) tempDir) {
    ...
}
```

#### Example with Custom Parent Directory

```java
...
import java.nio.file.Files;
...

class CustomParentFactory implements TempDirFactory {

    @Override
    public Path createTempDirectory(ExtensionContext context) throws IOException {
        return Files.createTempDirectory(Paths.get("/tmp-dir-root"), "junit");
    }

}
```

#### Example with Custom File System

Using [marschall/memoryfilesystem](https://github.com/marschall/memoryfilesystem):

```java
...
import java.nio.file.Files;
...

class MemoryFileSystemFactory implements TempDirFactory {

    private FileSystem fileSystem;

    @Override
    public Path createTempDirectory(ExtensionContext context) throws IOException {
        fileSystem = MemoryFileSystemBuilder.newEmpty().build();
        return Files.createTempDirectory(fileSystem.getPath("/"), "junit");
    }

    @Override
    public void close() throws IOException {
        fileSystem.close();
    }
}
```

Using [google/jimfs](https://github.com/google/jimfs):

```java
...
import java.nio.file.Files;
...

class JimfsFactory implements TempDirFactory {

    private FileSystem fileSystem;

    @Override
    public Path createTempDirectory(ExtensionContext context) throws IOException {
        FileSystem fileSystem = Jimfs.newFileSystem(Configuration.unix());
        return Files.createTempDirectory(fileSystem.getPath("/"), "junit");
    }

    @Override
    public void close() throws IOException {
        fileSystem.close();
    }
}
```

I would be happy to receive any feedback on the direction I have taken.

If this feature gets accepted, I would like to add a global configuration parameter like `junit.jupiter.tempdir.factory.default` in a separate PR.

### Open points
- [x] Behavior in case of `junit.jupiter.tempdir.scope=PER_CONTEXT`: the extension fails if a custom factory is set
- [x] Both in-memory file system libraries currently fail due to an unsupported `Path::toFile` call:
```
java.lang.UnsupportedOperationException: memory file system does not support #toFile()
	at com.github.marschall.memoryfilesystem.AbstractPath.toFile(AbstractPath.java:74)
	at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath.resetPermissions(TempDirectory.java:353)
	at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath.deleteAllFilesAndDirectories(TempDirectory.java:282)
	at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath.close(TempDirectory.java:268)
```

```
java.lang.UnsupportedOperationException
	at com.google.common.jimfs.JimfsPath.toFile(JimfsPath.java:387)
	at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath.resetPermissions(TempDirectory.java:353)
	at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath.deleteAllFilesAndDirectories(TempDirectory.java:282)
	at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath.close(TempDirectory.java:268)
```

To solve this, I catched the `UnsupportedOperationException` in `TempDirectory.CloseablePath::resetPermissions` and ignored it (b28450594c576d0433801fe73c0b9d377ea893a9). In support of this, the `Path::toFile` [javadoc](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Path.html#toFile--) mentions that the exception can be thrown if the `Path` is not associated with the default file system, which can be a potential use case once this SPI is offered.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Comment by scordio** [comment]
> Both in-memory file system libraries currently fail due to an unsupported `Path::toFile` call

After b28450594c576d0433801fe73c0b9d377ea893a9, both in-memory file system libraries work correctly.

**Comment by lacasseio** [comment]
I like what you did with this PR. I wonder if passing `prefix` is necessary given that all new examples ignore it and the standard implementation always uses the same prefix.

I also see this PR as a good way to customize the behaviour of the temporary directory, such as including a space. We have a custom test directory provider that namespace the directory per-class/method, i.e. `test-files/Foo/myTestMethod`, and the idea in this PR could help with that. However, we would need access to some information from the `ExecutionContext`, mainly the test method and/or test class. I won't say to include this feature in this PR, but I think it adds additional use cases to the `TempDirFactory` design. I don't think passing the `ExecutionContext` instead of the `prefix` is wise, but maybe something like `TempDirContext` could double for the `prefix` (if required). What do you think?

Finally, I think having a way to pass a global `TempDirFactory` implementation via system property would be useful. Users could still override the factory per `@TempDir` declaration. Maybe something like `junit.jupiter.tempdir.factory=com.example.MyTempDirFactory`.

I hope this PR gets a be more traction because it provides the last bit of customization that `TempDir` needs.

**Comment by scordio** [comment]
> I like what you did with this PR. 

@lacasseio thanks for the feedback! I have to admit I've put this topic a bit aside lately, there are still some points to clarify and test coverage to improve... I'll get it finalized sooner or later 🙂 

> I wonder if passing prefix is necessary given that all new examples ignore it and the standard implementation always uses the same prefix.

Indeed, the prefix doesn't seem so useful right now but I'd let the JUnit team judge what to do there.

> I also see this PR as a good way to customize the behaviour of the temporary directory, such as including a space. We have a custom test directory provider that namespace the directory per-class/method, i.e. `test-files/Foo/myTestMethod`, and the idea in this PR could help with that. However, we would need access to some information from the `ExecutionContext`, mainly the test method and/or test class. I won't say to include this feature in this PR, but I think it adds additional use cases to the `TempDirFactory` design. I don't think passing the `ExecutionContext` instead of the `prefix` is wise, but maybe something like `TempDirContext` could double for the `prefix` (if required). What do you think?

I agree `TempDirFactory` could have more input/control but I didn't feel the silver bullet yet, so I preferred starting with the smallest scope possible and for sure it could be extended with concrete use cases like yours. Especially here, I'd wait for some feedback from the core team before introducing something more complex.

> Finally, I think having a way to pass a global `TempDirFactory` implementation via system property would be useful. Users could still override the factory per `@TempDir` declaration. Maybe something like `junit.jupiter.tempdir.factory=com.example.MyTempDirFactory`.

That was exactly my idea, I've mentioned it in the middle of the description:

> If this feature gets accepted, I would like to add a global configuration parameter like `junit.jupiter.tempdir.factory` in a separate PR.

**Comment by lacasseio** [comment]
Looking at implementing something similar as part of my current "TempDir" extension, I think it would be important to ensure  `Factory` can be `AutoCloseable` as, for example, `Jimfs` are `AutoCloseable` which signals that the file system should be closed once created (potentially after a test).

**Comment by scordio** [comment]
Thanks for your feedback, @lacasseio! I need to update this topic and will check how to apply what you mentioned, as what the factory produces is already wrapped in a `CloseablePath`.

**Comment by lacasseio** [comment]
Great, I just noticed the example didn't close the FileSystem object and though about mentioning it for completeness. You are right that it won't change the current architecture.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+	/**
+	 * Return the {@code TempDirFactory} instance corresponding to the
+	 * given {@code Class}.
+	 *
+	 * @param factoryClass the factory's {@code Class}; never {@code null},
+	 * has to be a {@code TempDirFactory} implementation
+	 * @return a {@code TempDirFactory} implementation instance
+	 */
+	static TempDirFactory getTempDirFactory(Class<?> factoryClass) {
```
Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java:84]: Please move this to `TempDirectory` since it's an implementation detail.

**Comment by marcphilipp** [comment]
**Team Decision:** Check whether `TempDirFactory` implements `AutoCloseable` and, if so, close it in `CloseablePath` after deleting the temp dir.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java (hunk 1 above)
Comment (scordio) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java:84]: Here I mimicked what `DisplayNameGenerator` has today:

https://github.com/junit-team/junit5/blob/ac8a5bd6c70244a0a5ba8073469452d20af3f671/junit-jupiter-api/src/main/java/org/junit/jupiter/api/DisplayNameGenerator.java#L365-L382

Happy to take a different direction.

**Comment by olamy** [comment]
very good idea! This would be fantastic to have this available soon.
A global configuration would be good as well such `junit.jupiter.execution.tmpdir.factory` or any similar name.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	/**
+	 * Create a new temporary directory, using the given prefix to generate its name.
+	 * Depending on the implementation, the resulting {@code Path} may or may not be
+	 * associated with the default {@code FileSystem}.
+	 *
+	 * @param prefix the prefix string that can be used in generating the directory's name; never {@code null} or blank
+	 * @return the path to the newly created directory that did not exist before this method was invoked; never {@code null}
+	 * @throws Exception in case of failures
+	 */
+	Path createTempDirectory(String prefix) throws Exception;
```
Comment (olamy) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java:54]: could we have something such the current test class name in order to create directory depending on the name.
Use case is to debug content of created temp directories and/or files in case of test failures.

Comment (olamy) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java:52]: @scordio do you need a PR for this enhancement

Comment (scordio) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java:52]: Hi @olamy, I'm slowly catching up with the open points I have in this PR.

I'm ok with your proposal. I would also wait for feedback from the team before going in this direction as it requires a bit of refactoring to get the context as a method parameter. Once we get it, I'm happy to handle those changes.

Comment (scordio) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java]: Triggered by https://github.com/junit-team/junit5/pull/2958#issuecomment-145341381, I replaced the `prefix` parameter of `createTempDirectory` with the `ExtensionContext`.

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java [hunk 3]
```diff
@@ -250,6 +281,19 @@ public void close() throws IOException {
...
 			}
+
+			closeFactory();
+		}
+
+		private void closeFactory() throws IOException {
+			try {
+				if (factory instanceof AutoCloseable) {
+					((AutoCloseable) factory).close();
+				}
```
Comment (scordio) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java:292]: Would it make sense for `TempDirFactory` to extend `AutoCloseable` with an empty default implementation of `close()`?

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java [hunk 4]
```diff
@@ -330,12 +380,16 @@ private void resetPermissionsAndTryToDeleteAgain(Path path, IOException exceptio
...
+		private static void tryToResetPermissions(Path path) {
+			try {
+				File file = path.toFile();
+				file.setReadable(true);
+				file.setWritable(true);
+				if (Files.isDirectory(path)) {
+					file.setExecutable(true);
+				}
+			}
+			catch (UnsupportedOperationException ignore) {
```
Comment (scordio) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java:392]: I couldn't find an easy way to test this change.

I tried to create a test with an anonymous subtype of `Path` where `toFile` throws an `UnsupportedOperationException` and all the other methods are delegated to a `Path` created with `Files.createTempDirectory`, but the subtype causes failures before reaching this part of the code where `sun.nio.fs.UnixPath` is expected (I'm on Mac OS).

Any suggestions or should I leave this uncovered?

Another option could be to have one of the in-memory file system libraries as a test dependency and use it to cover this change. Would that make sense?

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+@API(status = EXPERIMENTAL, since = "5.10")
+public interface TempDirFactory {
+
+	/**
+	 * Create a new temporary directory, using the given prefix to generate its name.
+	 * Depending on the implementation, the resulting {@code Path} may or may not be
+	 * associated with the default {@code FileSystem}.
+	 *
+	 * @param context the current extension context; never {@code null}
+	 * @return the path to the newly created directory that did not exist before this method was invoked; never {@code null}
```
Comment (scordio) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java:50]: Similar to #2969, should we allow returning `null` for falling back to the default factory?

**Comment by scordio** [comment]
@marcphilipp @sormuras may I ask for an initial review?

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java [hunk 6]
```diff
@@ -332,12 +380,16 @@ private void resetPermissionsAndTryToDeleteAgain(Path path, IOException exceptio
...
 			}
```
Comment (marcphilipp) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java:393]: I think the `try-catch` block should only be around `path.toFile()` and we should add a comment to the catch block explaining the well-known cases when this happens.

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java [hunk 7]
```diff
@@ -187,13 +209,22 @@ private void assertSupportedType(String target, Class<?> type) {
...
-	private Object getPathOrFile(AnnotatedElement sourceElement, Class<?> type, CleanupMode cleanupMode,
-			ExtensionContext extensionContext) {
-		Namespace namespace = getScope(extensionContext) == Scope.PER_DECLARATION //
+	@SuppressWarnings("deprecation")
+	private Object getPathOrFile(AnnotatedElement sourceElement, Class<?> type, TempDirFactory factory,
+			CleanupMode cleanupMode, ExtensionContext extensionContext) {
+		Scope scope = getScope(extensionContext);
+
+		if (scope == Scope.PER_CONTEXT && factory != TempDirFactory.Standard.INSTANCE) {
+			throw new ExtensionConfigurationException("Custom @TempDir factory is not supported with "
```
Comment (marcphilipp) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java:218]: Why is it not supported with that scope?

**Review by marcphilipp — CHANGES_REQUESTED** [review]
Looks very promising! Could you please add an integration test using jimfs or memoryfilesystem by introducing an additional test dependency?

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java (hunk 7 above)
Comment (scordio) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java:218]: It's basically connected to your comment at https://github.com/junit-team/junit5/issues/2400#issuecomment-1096279042 (I guess).

For example, if `@TempDir` is declared on two fields in the same test class but each annotation has its own factory, I assume the PER_CONTEXT scope would provide the same instance to both, making one of the factories useless. I haven't demonstrated this with a test case yet, but it's something I can try 👍 

If my guess is correct, a local factory configuration should be disallowed due to unpredictable behavior, but a global configuration via the `junit.jupiter.tempdir.factory.default` configuration property (not part of this PR but I'm already working on the changes) could be allowed.

Or do you see it differently?

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java (hunk 6 above)
Comment (scordio) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java]: Done in 24b6a0b.

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java (hunk 4 above)
Comment (scordio) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java]: Solved in 24b6a0b with integration tests for both MemoryFileSystem and Jimfs.

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java (hunk 3 above)
Comment (scordio) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java:297]: Triggered by the integration tests added in 24b6a0b5, I went ahead with this change. Please, let me know if you see anything wrong with it.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+	 * @throws Exception in case of failures
+	 */
+	Path createTempDirectory(ExtensionContext context) throws Exception;
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	default void close() throws IOException {
+	}
```
Comment (scordio) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java:61]: What's the policy for documenting default overridden methods? Is `{@inheritDoc}` enough, should I leave it empty, or should I write a dedicated Javadoc?

Looking around, I see that `ResourceLock::close` has no Javadoc defined:

https://github.com/junit-team/junit5/blob/6b12a22dd30bfcb9012be456895cad8df1ea9366/junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/ResourceLock.java#L41-L44

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java (hunk 7 above)
Comment (scordio) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java:218]: @marcphilipp I was able to confirm my assumption via this example in `TempDirectoryPerContextTests` (without the restriction in `TempDirectory`):

multiple_temp-dir-factories

If you check the path injected into `tempDir2` (shown in gray by the debugger), it's the one created by `Factory1` which I wouldn't consider to be the expected behavior.

For this reason, I would keep the restriction. What do you think?

Comment (marcphilipp) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java:218]: Yes, that makes sense. 👍

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java (hunk 8 above)
Comment (marcphilipp) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java:61]: We don't have a documented guideline for this case. I think using `{@inheritDoc}` is fine here.

File: junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/extension/TempDirectoryPerDeclarationTests.java [hunk 9]
```diff
@@ -1131,4 +1150,46 @@ public Path createTempDirectory(ExtensionContext context) throws Exception {
...
+				return Files.createTempDirectory(fileSystem.getPath("/"), "prefix");
+			}
+		}
+
+	}
+
+	static class FactoryWitJimfsTestCase {
+
+		@Test
+		void test(@TempDir(factory = Factory.class) Path tempDir) {
```
Comment (marcphilipp) [inline junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/extension/TempDirectoryPerDeclarationTests.java:1182]: What should happen in the following case:
```java
@Test
void test(@TempDir(factory = Factory.class) Path a, @TempDir(factory = Factory.class) Path b)
```

Without specifying `factory`, two different paths from the same (default) file system will be injected. With `factory` they will be from different file systems. I wonder if we need to support that use case somehow. 🤔

**Review by marcphilipp — APPROVED** [review]

File: junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/extension/TempDirectoryPerDeclarationTests.java (hunk 9 above)
Comment (scordio) [reply junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/extension/TempDirectoryPerDeclarationTests.java:1182]: Not sure I understood your concern.

In case the same factory is specified on both annotations, how the underlying file system is handled should be an implementation detail of the factory.

Maybe you're referring to the fact that each annotation will instantiate a different factory. If something like the file system has to be shared, it should be declared as static (with additional care in the `close()` method).

Is that what you were thinking?

The Javadoc could maybe highlight such cases. What do you think?

Comment (marcphilipp) [reply junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/extension/TempDirectoryPerDeclarationTests.java:1182]: Yes, I was wondering if there's going to be a need to use the same `FileSystem`. However, one could even solve this by using the `ExtensionContext` and putting a `CloseableResource` in the root `Store`. For now, I don't think we have to document anything in particular given we have no real-world use case.

**Comment by marcphilipp** [comment]
@scordio This PR is a great next step but I don't think it solves #1786, does it?

**Comment by scordio** [comment]
> @scordio This PR is a great next step but I don't think it solves #1786, does it?

You're right, I guess I had something else in mind when I mentioned it. I take it out from the description.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java (hunk 5 above)
Comment (scordio) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java:51]: @marcphilipp What do you think of this open point? Should I leave it aside, to be added only if there is a concrete request?

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc [hunk 10]
```diff
@@ -2621,5 +2621,19 @@ The default cleanup mode is `ALWAYS`. You can use the
...
+
+Factories can be created by implementing `TempDirFactory`. The default implementation
+available in Jupiter delegates the directory creation to
+`java.nio.file.Files::createTempDirectory`, passing `junit` as the prefix string
+to be used in generating the directory's name.
+
+[source,java,indent=0]
+.A test class with a temporary directory created by a custom factory
+----
+include::{testDir}/example/TempDirectoryDemo.java[tags=user_guide_factory]
```
Comment (scordio) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:2638]: Not sure if this is enough or if there should be more. For example, should I mention `close()` for managing underlying resources like a custom `FileSystem`?

Would it make sense to add an example with an in-memory file system?

(If yes, any preference between [google/jimfs](https://github.com/google/jimfs) and [marschall/memoryfilesystem](https://github.com/marschall/memoryfilesystem)?)

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java (hunk 5 above)
Comment (marcphilipp) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java:51]: I don't think we should allow that.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc (hunk 10 above)
Comment (marcphilipp) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:2638]: Yes, I think that would make a better example. Maybe use jimfs because it seems more popular?

Comment (scordio) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: Done in d92f151.

**Comment by scordio** [comment]
@marcphilipp I think I'm done with my changes, all the open points should be covered. Please, let me know if there is anything to adjust.

**Comment by scordio** [comment]
> If this feature gets accepted, I would like to add a global configuration parameter like `junit.jupiter.tempdir.factory.default` in a separate PR.

Initial draft at https://github.com/scordio/junit5/compare/2400_temp_dir_factory...scordio:junit5:gh-2400_temp_dir_factory_config_property.

I'll raise a PR once this one is merged.

File: documentation/documentation.gradle.kts [hunk 11]
```diff
@@ -54,6 +54,10 @@ dependencies {
 		because("ApiReportGenerator needs it")
 	}
 
+	testImplementation(libs.jimfs) {
+		because("Jimfs is used in src/main/test")
```
Comment (marcphilipp) [inline documentation/documentation.gradle.kts:58]:
```suggestion
		because("Jimfs is used in src/test/java")
```

**Review by marcphilipp — APPROVED** [review]

File: documentation/documentation.gradle.kts (hunk 11 above)
Comment (scordio) [reply documentation/documentation.gradle.kts]: Whoops, good catch 😅 thanks!

**Comment by marcphilipp** [comment]
@scordio Thanks a lot for the productive collaboration on this one! 🎉
